File: sutra_ws/src/sutra_gnc/ground_station/src/engine/core/missionStateMachine.ts
import type { MissionState } from '../types';

type StateListener = (state: MissionState, previousState: MissionState, reason?: string) => void;

/**
 * Valid state transitions mapping
 */
const VALID_TRANSITIONS: Record<MissionState, MissionState[]> = {
  IDLE: ['PLANNING', 'EMERGENCY'],
  PLANNING: ['READY', 'IDLE', 'ABORTED', 'EMERGENCY'],
  READY: ['UPLOADING', 'PLANNING', 'ABORTED', 'EMERGENCY'],
  UPLOADING: ['ARMING', 'READY', 'ABORTED', 'EMERGENCY'],
  ARMING: ['TAKEOFF', 'READY', 'ABORTED', 'EMERGENCY'],
  TAKEOFF: ['MISSION', 'HOLD', 'RTL', 'LANDING', 'ABORTED', 'EMERGENCY'],
  MISSION: ['HOLD', 'RTL', 'LANDING', 'COMPLETE', 'ABORTED', 'EMERGENCY'],
  HOLD: ['MISSION', 'RTL', 'LANDING', 'ABORTED', 'EMERGENCY'],
  RTL: ['LANDING', 'HOLD', 'ABORTED', 'EMERGENCY'],
  LANDING: ['COMPLETE', 'RTL', 'ABORTED', 'EMERGENCY'],
  COMPLETE: ['IDLE', 'PLANNING'],
  ABORTED: ['IDLE', 'PLANNING', 'RTL', 'LANDING'],
  EMERGENCY: ['LANDING', 'RTL', 'IDLE', 'ABORTED']
};

export class MissionStateMachine {
  private currentState: MissionState = 'IDLE';
  private listeners: Set<StateListener> = new Set();
  private history: { state: MissionState; timestamp: string; reason?: string }[] = [];

  constructor(initialState: MissionState = 'IDLE') {
    this.currentState = initialState;
    this.recordHistory(initialState, 'Initialization');
  }

  public getState(): MissionState {
    return this.currentState;
  }

  public canTransitionTo(nextState: MissionState): boolean {
    const allowed = VALID_TRANSITIONS[this.currentState] || [];
    return allowed.includes(nextState);
  }

  public transitionTo(nextState: MissionState, reason?: string): boolean {
    if (this.currentState === nextState) return true;

    if (!this.canTransitionTo(nextState)) {
      console.warn(`[MissionStateMachine] Invalid transition from ${this.currentState} to ${nextState}`);
      return false;
    }

    const previousState = this.currentState;
    this.currentState = nextState;
    this.recordHistory(nextState, reason);

    this.listeners.forEach((listener) => listener(nextState, previousState, reason));
    return true;
  }

  public forceState(nextState: MissionState, reason?: string): void {
    const previousState = this.currentState;
    this.currentState = nextState;
    this.recordHistory(nextState, `FORCED: ${reason || ''}`);
    this.listeners.forEach((listener) => listener(nextState, previousState, reason));
  }

  public subscribe(listener: StateListener): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  public getHistory() {
    return [...this.history];
  }

  private recordHistory(state: MissionState, reason?: string) {
    this.history.push({
      state,
      timestamp: new Date().toISOString(),
      reason
    });
  }
}

export const missionStateMachine = new MissionStateMachine();

File: sutra_ws/src/sutra_gcs/engine/mission_state_machine.py
"""
Smart Horizon GCS — Mission Execution Finite State Machine (FSM)
Subsystem: Mission Engine (Phase 5)
"""

import logging
from dataclasses import replace
from typing import Dict, List, Optional, Set

from services.event_bus import EventBus, get_event_bus
from services.logging_service import get_logger
from state.application_state import StateStore, get_state_store
from state.mission_state import MissionStateEnum

logger = logging.getLogger("sutra_gcs.mission_state_machine")


class MissionStateMachine:
    """
    Strict deterministic Finite State Machine governing autonomous flight plan lifecycle.
    Prevents illegal state transitions, enforces pre-flight validation gates,
    and synchronizes mission state across the application.
    """

    # Explicit whitelist of allowed directional state transitions
    ALLOWED_TRANSITIONS: Dict[MissionStateEnum, Set[MissionStateEnum]] = {
        MissionStateEnum.IDLE: {
            MissionStateEnum.PLANNING,
            MissionStateEnum.VALIDATING,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.PLANNING: {
            MissionStateEnum.VALIDATING,
            MissionStateEnum.IDLE,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.VALIDATING: {
            MissionStateEnum.READY,
            MissionStateEnum.PLANNING,
            MissionStateEnum.IDLE,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.READY: {
            MissionStateEnum.ARMING,
            MissionStateEnum.PLANNING,
            MissionStateEnum.UPLOADING,
            MissionStateEnum.IDLE,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.UPLOADING: {
            MissionStateEnum.ARMING,
            MissionStateEnum.READY,
            MissionStateEnum.ABORTED,
        },
        MissionStateEnum.ARMING: {
            MissionStateEnum.TAKEOFF,
            MissionStateEnum.READY,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.TAKEOFF: {
            MissionStateEnum.MISSION,
            MissionStateEnum.HOLD,
            MissionStateEnum.RTL,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.MISSION: {
            MissionStateEnum.HOLD,
            MissionStateEnum.RTL,
            MissionStateEnum.LANDING,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.HOLD: {
            MissionStateEnum.MISSION,
            MissionStateEnum.RTL,
            MissionStateEnum.LANDING,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.RTL: {
            MissionStateEnum.LANDING,
            MissionStateEnum.HOLD,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.LANDING: {
            MissionStateEnum.COMPLETE,
            MissionStateEnum.ABORTED,
            MissionStateEnum.EMERGENCY,
        },
        MissionStateEnum.COMPLETE: {
            MissionStateEnum.IDLE,
            MissionStateEnum.PLANNING,
        },
        MissionStateEnum.ABORTED: {
            MissionStateEnum.IDLE,
            MissionStateEnum.PLANNING,
        },
        MissionStateEnum.EMERGENCY: {
            MissionStateEnum.IDLE,
            MissionStateEnum.PLANNING,
            MissionStateEnum.ABORTED,
        },
    }

    def __init__(
        self,
        state_store: Optional[StateStore] = None,
        event_bus: Optional[EventBus] = None,
    ) -> None:
        self.state_store = state_store or get_state_store()
        self.event_bus = event_bus or get_event_bus()
        self.logger = get_logger("state_machine")

    @property
    def current_state(self) -> MissionStateEnum:
        """Returns active mission state from StateStore."""
        return self.state_store.get_state().mission_state.state

    def can_transition_to(self, target_state: MissionStateEnum) -> bool:
        """Checks if a transition from current_state to target_state is permitted."""
        curr = self.current_state
        allowed = self.ALLOWED_TRANSITIONS.get(curr, set())
        return target_state in allowed

    def transition_to(self, target_state: MissionStateEnum, reason: str = "") -> bool:
        """
        Attempts to advance the FSM to target_state.
        Returns True if successful, False if rejected.
        """
        curr = self.current_state
        if not self.can_transition_to(target_state):
            self.logger.warning(
                f"REJECTED illegal FSM transition: {curr.value} -> {target_state.value} ({reason})"
            )
            return False

        self.logger.info(
            f"FSM Transition: {curr.value} -> {target_state.value} [{reason or 'Nominal'}]"
        )

        # Update StateStore
        self.state_store.update_state(
            lambda s: replace(
                s,
                mission_state=replace(s.mission_state, state=target_state),
            )
        )

        # Emit EventBus event
        self.event_bus.emit(
            f"mission.{target_state.value.lower()}",
            payload={
                "previous_state": curr.value,
                "current_state": target_state.value,
                "reason": reason,
            },
            source="mission_state_machine",
        )
        return True

    def reset(self) -> None:
        """Resets the state machine to IDLE."""
        self.state_store.update_state(
            lambda s: replace(
                s,
                mission_state=replace(s.mission_state, state=MissionStateEnum.IDLE),
            )
        )
        self.event_bus.emit(
            "mission.idle",
            payload={"current_state": "IDLE", "reason": "FSM Reset"},
            source="mission_state_machine",
        )
